src/main/java/network/brightspots/rcv/CommonDataFormatReader.java

```diff
@@ -118,15 +140,209 @@
+    for (ContestSelection contestSelection : contestToTabulate.ContestSelection) {
+      if (contestSelection.CandidateIds.length > 1) {
+        Logger.warning(
+            "Multiple CandidateIds found in ContestSelection \"%s\". "
+                + " Only the first one will be processed.",
+            contestSelection.ObjectId);
+      }
+      contestSelectionById.put(contestSelection.ObjectId, contestSelection);
+    }
+
+    // build a map of GpUnits (aka precinct or district)
+    HashMap<String, GpUnit> gpUnitById = new HashMap<>();
+    for (GpUnit gpUnit : cvrReport.GpUnit) {
+      gpUnitById.put(gpUnit.ObjectId, gpUnit);
+    }
+
+    // process the Cvrs
+    for (CVR cvr : cvrReport.CVR) {
+      CVRContest contest = getCvrContest(cvr, contestToTabulate);
+      if (contest == null) {
+        // the CVR does not contain any votes for this contest
+        continue;
+      }
+      List<Pair<Integer, String>> rankings = new ArrayList<>();
+      // parse CVRContestSelections into rankings
+      // they will be null for an undervote
+      if (contest.CVRContestSelection != null) {
+        for (CVRContestSelection cvrContestSelection : contest.CVRContestSelection) {
+          if (cvrContestSelection.Status != null && cvrContestSelection.Status
+              .equals(STATUS_NEEDS_ADJUDICATION)) {
+            Logger.info("Contest Selection needs adjudication.  Skipping.");
+            continue;
+          }
+          String contestSelectionId = cvrContestSelection.ContestSelectionId;
+          ContestSelection contestSelection = contestSelectionById
+              .get(contestSelectionId);
+          if (contestSelection == null) {
+            Logger.severe("ContestSelection \"%s\" from CVR not found!", contestSelectionId);
+            throw new CvrParseException();
+          }
+          String candidateId;
+          // check for declared write-in:
+          if (contestSelection.IsWriteIn != null && contestSelection.IsWriteIn
+              .equals(BOOLEAN_TRUE)) {
+            candidateId = this.config.getUndeclaredWriteInLabel();
+          } else {
+            Candidate candidate = candidateById
+                .get(contestSelection.CandidateIds[0]);
+            if (candidate == null) {
+              Logger.severe("CandidateId \"%s\" from ContestSelectionId \"%s\" not found!",
+                  contestSelection.CandidateIds[0], contestSelection.ObjectId);
+              throw new CvrParseException();
+            }
+            candidateId = candidate.Name;
+            if (candidateId.equals(config.getOvervoteLabel())) {
+              candidateId = Tabulator.EXPLICIT_OVERVOTE_LABEL;
+            }
+            if (!config.getCandidateCodeList().contains(candidateId)) {
+              Logger
+                  .severe(
+                      "Contest Selection CandidateId: \"%s\" from CVR is not in the config file!",
+                      candidateId);
+              throw new CvrParseException();
```

Comment:
I'm still wondering whether we should be using unrecognizedCandidateCounts here instead of throwing exceptions.

Reply:
@tarheel I agree, in fact I already made that change in both xml and json parsers in the other open PR #506 which builds on this code.